core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala

```diff
@@ -44,11 +45,12 @@ class H2ODataFrame[T <: water.fvec.Frame](@transient val frame: T,
   def this(@transient frame: T)
           (@transient sc: SparkContext) = this(frame, null)(sc)
 
+  val typesAll: Array[DataType] = frame.vecs map ReflectionUtils.dataTypeFor
 
-  val typesAll: Array[DataType] = frame.vecs().indices.map(idx => vecTypeToDataType(frame.vec(idx))).toArray
   /** Create new types list which describes expected types in a way external H2O backend can use it. This list
     * contains types in a format same for H2ODataFrame and H2ORDD */
-  val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, typesAll)
+  lazy val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, typesAll)
```

Comment:
I am not sure at all whether this stuff will work. Laziness can be kicked
out, it's not an issue; as soon as the thing serializes. Since there are no
tests for that, we don't know for sure how it behaves.

Thanks,
-Vlad

On Thu, Sep 8, 2016 at 1:11 AM, Jakub Háva <notifications@github.com> wrote:

> In core/src/main/scala/org/apache/spark/h2o/converters/H2ODataFrame.scala
> <https://github.com/h2oai/sparkling-water/pull/76#discussion_r77962509>:
>
> >    /** Create new types list which describes expected types in a way external H2O backend can use it. This list
> >      * contains types in a format same for H2ODataFrame and H2ORDD */
> > -  val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, typesAll)
> > +  lazy val expectedTypesAll: Option[Array[Byte]] = ConverterUtils.prepareExpectedTypes(isExternalBackend, typesAll)
>
> I'm not sure how lazy behaves in distributed environments. Spark
> serialises the outer scope of compute method, but not sure how it does
> with lazy val.
>
> Will this work ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/h2oai/sparkling-water/pull/76/files/1db8264c73784cd33bda4a3a9ab9eefc8e55e4fd#r77962509>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAOuK1D9MpvDhOG-VvagyODxaw9o5QUMks5qn8M3gaJpZM4J3eDi>
> .
>